pkg/connector/service-principals.go

```diff
@@ -207,6 +200,177 @@ func (s *servicePrincipalBuilder) Grants(ctx context.Context, resource *v2.Resou
 	return rv, "", nil, nil
 }
 
+func (s *servicePrincipalBuilder) Grant(ctx context.Context, principal *v2.Resource, entitlement *v2.Entitlement) (annotations.Annotations, error) {
+	l := ctxzap.Extract(ctx)
+
+	if !isValidPrincipal(principal.Id) {
+		l.Warn(
+			"databricks-connector: only users, groups and service principals can be granted service principal permissions",
+			zap.String("principal_id", principal.Id.String()),
+			zap.String("principal_type", principal.Id.ResourceType),
+		)
+
+		return nil, fmt.Errorf("databricks-connector: only users, groups and service principals can be granted service principal permissions")
+	}
+
+	groupTrait, err := rs.GetGroupTrait(entitlement.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to get group trait: %w", err)
+	}
+
+	parentType, parentID, err := getParentInfoFromProfile(groupTrait.Profile)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to get parent info from group profile: %w", err)
+	}
+
+	if parentType == workspaceResourceType.Id {
+		s.client.SetWorkspaceConfig(parentID)
+	} else {
+		s.client.SetAccountConfig()
+	}
+
+	applicationId, ok := rs.GetProfileStringValue(groupTrait.Profile, "application_id")
+	if !ok {
+		return nil, fmt.Errorf("databricks-connector: failed to get application_id from service principal profile")
+	}
+
+	ruleSets, err := s.client.ListRuleSets(ctx, ServicePrincipalsType, applicationId)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to list rule sets for service principal %s (%s): %w", principal.Id.Resource, applicationId, err)
+	}
+
+	principalID, err := preparePrincipalID(ctx, s.client, principal.Id.ResourceType, principal.Id.Resource)
+	if err != nil {
+		return nil, fmt.Errorf("databricks-connector: failed to prepare principal id: %w", err)
+	}
+
+	found := false
+
+	for i, ruleSet := range ruleSets {
```

Comment:
This code is so close to being the same in three places, but it's just different enough that I don't see a way to replace it with one helper function. Also it'd have to return a flag that lets the caller know the entitlement is already granted, so the caller can return early. I guess slightly different code is the best of the available options.